File: sdk/samples/FrameworkSamples/NETCF/Applications/RomanLegion/vb/romanlegion/baseobj.vb
'
' File: BaseObj.vb
'
'---------------------------------------------------------------------
' This file is part of the Microsoft .NET Framework SDK Code Samples.
'
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
' This source code is intended only as a supplement to Microsoft
' Development Tools and/or on-line documentation.  See these other
' materials for detailed information regarding Microsoft code samples.
' 
' THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
' KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
' IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
' PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Namespace Microsoft.Samples.RomanLegion

   Public Class BaseObj
   
       ' Image index to draw
       Private myActiveImage As Integer = 0
   
       ' Object's Left
       Private myLeft As Integer
       ' Object's Top
       Private myTop As Integer
   
       ' Object's Left before move
       Private myLeftOld As Integer
       ' Object's Top before move
       Private myTopOld As Integer
   
       ' Object's Speed to right 
       Private myHorizontalVelocity As Integer
       ' Object's Speed to down
       Private myVerticalVelocity As Integer
   
       ' Object's Width 
       Private myWidth As Integer
       ' Object's Height
       Private myHeight As Integer
   
       ' Max x Position
       Private myMaximumX As Integer
       ' Max y Position
       Private myMaximumY As Integer
   
       ' Current game
       Private myCurrentGame As Game
   
       ' is to draw ?
       Private fActive As Boolean
   
       ' Object Bitmap
       Private myCells() As Bitmap
   
       ' Object Attributes (which color range is transparent ?)
       Private myCellAttributes() As Imaging.ImageAttributes
   
        Protected Property LeftOld As Integer
            Get
                Return myLeftOld
            End Get
            Set
                myLeftOld = Value
            End Set
        End Property

        Protected Property TopOld As Integer
            Get
                Return myTopOld
            End Get
            Set
                myTopOld = Value
            End Set
        End Property

        Protected Property HorizontalVelocity As Integer
            Get
                Return myHorizontalVelocity 
            End Get
            Set
                myHorizontalVelocity = Value
            End Set
        End Property

        Protected Property VerticalVelocity As Integer
            Get
                Return myVerticalVelocity
            End Get
            Set
                myVerticalVelocity = Value
            End Set
        End Property

        Protected Property MaximumX As Integer
            Get
                Return myMaximumX 
            End Get
            Set
                myMaximumX = Value
            End Set
        End Property

        Protected Property MaximumY As Integer
            Get
                Return myMaximumY
            End Get
            Set
                myMaximumY = Value
            End Set
        End Property

        Protected Property CurrentGame As Game
            Get
                Return myCurrentGame
            End Get
            Set
                myCurrentGame = Value
            End Set
        End Property

        Protected Property AnimationCells As Bitmap()
            Get
                Return myCells
            End Get
            Set
                myCells = Value
            End Set
        End Property

        Protected Property CellAttributes As System.Drawing.Imaging.ImageAttributes()
            Get
                Return myCellAttributes
            End Get
            Set
                myCellAttributes = Value
            End Set
        End Property

       Public Sub New(ByVal game As Game)
   
           myCurrentGame = game
   
           myMaximumX = myCurrentGame.WorldBounds.Width
           myMaximumY = myCurrentGame.WorldBounds.Height
   
           myLeftOld = 0
           myTopOld = 0
       End Sub
   
       Public Sub Draw(ByVal gx As Graphics)
   
           ' if is to draw
           If (fActive) Then
   
               gx.DrawImage( _
                   myCells(myActiveImage), _
                   New Rectangle(myLeft, myTop, myWidth, myHeight), _
                   0, _
                   0, _
                   myWidth, _
                   myHeight, _
                   GraphicsUnit.Pixel, _
                   myCellAttributes(myActiveImage))
           End If
       End Sub
   
       Public Sub Load(ByVal bmpName As String, ByVal bitmapIndex As Integer)
   
           Try
   
               ' Load BMP from file
                myCells(bitmapIndex) = New Bitmap(myCurrentGame.ExecutablePath + "Content\" + bmpName)
               myWidth = myCells(bitmapIndex).Width
               myHeight = myCells(bitmapIndex).Height
   
           Catch
   
               ' Any error reading the BMP, create a dummy one
               myCells(bitmapIndex) = New Bitmap(myWidth, myHeight)
               Graphics.FromImage(myCells(bitmapIndex)).FillEllipse(New SolidBrush(Color.White), New Rectangle(0, 0, myWidth - 1, myHeight - 1))
           End Try
   
           ' Get the color of a background pixel as the Pixel 0,0 
           Dim TranspColor As Color = myCells(bitmapIndex).GetPixel(0, 0)
   
           ' Set the Attributes for the Transparent color
           myCellAttributes(bitmapIndex) = New Imaging.ImageAttributes()
           myCellAttributes(bitmapIndex).SetColorKey(TranspColor, TranspColor)
       End Sub
   
       Public Sub Restore()
   
           ' Restore the Game Background
           myCurrentGame.BufferGraphics.DrawImage( _
               myCurrentGame.BackBmp, _
               myLeft, _
               myTop, _
               New Rectangle(myLeft, myTop, myWidth, myHeight), _
               GraphicsUnit.Pixel)
       End Sub
   
       Public Property Bounds() As Rectangle
           Get
               Return (New Rectangle(myLeft, myTop, myWidth, myHeight))
           End Get
           Set(ByVal Value As Rectangle)
           End Set
       End Property
   
       Public Property DirtyArea() As Rectangle
           Get
               Return (Rectangle.Union(New Rectangle(myLeft, myTop, myWidth, myHeight), New Rectangle(myLeftOld, myTopOld, myWidth, myHeight)))
           End Get
           Set(ByVal Value As Rectangle)
           End Set
       End Property
   
       ' Expose object Draw property
       Public Property IsAlive As Boolean
           Get
               Return (fActive)
           End Get
           Set(ByVal Value As Boolean)
               fActive = Value
           End Set
       End Property
       
       Public Property CurrentImage As Integer
           Get
               Return (myActiveImage)
           End Get
           Set
               myActiveImage = Value
           End Set
       End Property
       
       Public Property CellLeft As Integer
           Get
               Return (myLeft)
           End Get
           Set
               myLeft = Value
           End Set
       End Property
       
       Public Property CellTop As Integer
           Get
               Return (myTop)
           End Get
           Set
               myTop = Value
           End Set
       End Property
       
       Public Property CellWidth As Integer
           Get
               Return (myWidth)
           End Get
           Set
               myWidth = Value
           End Set
       End Property
       
       Public Property CellHeight As Integer
           Get
               Return (myHeight)
           End Get
           Set
               myHeight = Value
           End Set
       End Property
       
   End Class

End Namespace
   

File: sdk/samples/FrameworkSamples/NETCF/Applications/RomanLegion/cs/romanlegion/baseobj.cs
//=====================================================================
//  File:      BaseObj.cs
//
//---------------------------------------------------------------------
//  This file is part of the Microsoft .NET Framework SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------
 
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.Windows.Forms;
using System.Reflection;
using System.Threading;
using System.ComponentModel;

namespace Microsoft.Samples.RomanLegion
{
	public class BaseObj
	{
		// Image index to draw
		private int activeImage = 0;

		// Object's Left
		private int left;
		// Object's Top
		private int top;
	
		// Object's Left before move
		private int leftOld;
		// Object's Top before move
		private int topOld;
	
		// Object's Speed to right 
		private int horizontalVelocity;
		// Object's Speed to down
		private int verticalVelocity;

		// Object's Width 
		private int width;
		// Object's Height
		private int height;
	
		// Max x Position
		private int maxX;
		// Max y Position
		private int maxY;
	
		// Current game
		private Game currentGame;
	
		// is to draw ?
		private bool fActive;

		// Object Bitmap
		private  Bitmap[] bitmapCells;
		
		// Object Attributes (which color range is transparent ?)
		private ImageAttributes[] cellAttributes;

        protected int LeftOld
        {
            get
            {
                return(leftOld);
            }
            set
            {
                leftOld = value;
            }
        }

        protected int TopOld
        {
            get
            {
                return(topOld);
            }
            set
            {
                topOld = value;
            }
        }

        protected int HorizontalVelocity
        {
            get
            {
                return(horizontalVelocity);
            }

            set
            {
                horizontalVelocity = value;
            }
        }

        protected int VerticalVelocity
        {
            get
            {
                return(verticalVelocity);
            }
            set
            {
                verticalVelocity = value;
            }
        }

        protected int MaximumX
        {
            get
            {
                return(maxX);
            }
            set
            {
                maxX = value;
            }
        }

        protected int MaximumY
        {
            get
            {
                return(maxY);
            }
            set
            {
                maxY = value;
            }
        }

        protected Game CurrentGame
        {
            get
            {
                return(currentGame);
            }
            set
            {
                currentGame = value;
            }
        }

        protected Bitmap[] AnimationCells
        {
            get
            {
                return(bitmapCells);
            }
            set
            {
                bitmapCells = value;
            }
        }

        protected ImageAttributes[] CellAttributes
        {
            get
            {
                return(cellAttributes);
            }
            set
            {
                cellAttributes = value;
            }
        }

		public BaseObj(Game game)
		{
			currentGame = game;

			maxX = currentGame.WorldBounds.Width;
			maxY = currentGame.WorldBounds.Height;

			leftOld = 0;
			topOld = 0;
		}

        public void Draw(Graphics gx)
		{
			// if is to draw
			if (fActive)
			{
				gx.DrawImage( 
					bitmapCells[activeImage],
					new Rectangle(left, top, width , height),
					0,
					0,
					width,
					height,
                    GraphicsUnit.Pixel,
					cellAttributes[activeImage]);
			}
		}

		public void Load(string bmpName, int bitmapIndex)
		{
			try
			{
				// Load BMP from file
				bitmapCells[bitmapIndex] = new Bitmap(
                    currentGame.BaseDirectory + @"Content\" + bmpName);
				width=bitmapCells[bitmapIndex].Width;
				height=bitmapCells[bitmapIndex].Height;
			}
			catch
			{
				// Any error reading the BMP, create a dummy one
				bitmapCells[bitmapIndex] = new Bitmap(width, height);
				Graphics.FromImage(bitmapCells[bitmapIndex]).FillEllipse(new SolidBrush(Color.White), new Rectangle(0, 0, width - 1, height - 1));
			}

			// Get the color of a background pixel as the Pixel 0,0 
			Color TranspColor = bitmapCells[bitmapIndex].GetPixel(0, 0);
			
			// Set the Attributes for the Transparent color
			cellAttributes[bitmapIndex] = new ImageAttributes();
			cellAttributes[bitmapIndex].SetColorKey(TranspColor, TranspColor);
		}

		public void Restore()
		{
			// Restore the Game Background
			currentGame.OffscreenGraphics.DrawImage(
				currentGame.BackBmp,
				left,
				top,
				new Rectangle(left, top, width , height),
				GraphicsUnit.Pixel);
		}

		public Rectangle Bounds
		{
			get
			{
				return(new Rectangle(left, top, width, height));
			}
		}

		public Rectangle DirtyArea
		{
			get
			{
				return(Rectangle.Union(new Rectangle(left, top, width, height), new Rectangle(leftOld, topOld, width, height)));
			}
		}

		// Expose object Draw property
		public bool IsAlive
		{
			get
			{
				return(fActive);
			}
			set
			{
				fActive = value;
			}
		}

        public int ActiveImage
        {
            get
            {
                return(activeImage);
            }
            set
            {
                activeImage = value;
            }
        }

        public int Left
        {
            get
            {
                return(left);
            }
            protected set
            {
                left = value;
            }
        }

        public int Top
        {
            get
            {
                return(top);
            }
            protected set
            {
                top = value;
            }
        }

        public int Width
        {
            get
            {
                return(width);
            }
            protected set
            {
                width = value;
            }
        }

        public int Height
        {
            get
            {
                return(height);
            }
            protected set
            {
                height = value;
            }
        }
	}
}
